# ДЗ #1   

**Что сделать?**  
Решить все задания и заполнить соответствующий гугл-опрос.  
Ссылка на гугл-опрос --> [тут](https://docs.google.com/forms/d/e/1FAIpQLSefdNkCvrhOQUFgrWfRZm4etiQYR5Ee9boieC4JlpZem9lu0w/viewform?usp=sf_link).  

##  Задача
Знакомство с библиотеками pandas, numpy.  
Для примера будем обрабатывать датасет твитов Д. Трампа.  
Названия полей в принципе интуитивно понятные, но на всякий случай вот описание:  

| Column        | Description   |
| :------------- |:-------------| 
| Date      | Дата твита | 
| Time      | Время твита      |
| Tweet_Text      | Текст твита | 
| Type      | Тип твита      |
| Hashtags      | Хештеги твита      |
| Tweet_Id      | Уникальный идентификатор твита | 
| Tweet_Url      | http-ссылка на твит      |
| Likes      | Количество лайков твита      |
| Retweets      | Количество ретвитов твита | 


Выполняйте задачи последовательно. Предыдущие этапы могут влиять на ответы для будущих задач.  
**подсказка:** *все основные примеры работы функций описаны в тренировочном файле к уроку.*

In [18]:
import pandas as pd
import numpy as np

df = pd.read_csv('datasets/DonaldTrumpTweets.csv')
df


Date      Time                                         Tweet_Text  \
0     16-11-11  15:26:37  Today we express our deepest gratitude to all ...   
1     16-11-11  13:33:35  Busy day planned in New York. Will soon be mak...   
2     16-11-11  11:14:20  Love the fact that the small groups of protest...   
3     16-11-11   2:19:44  Just had a very open and successful presidenti...   
4     16-11-11   2:10:46  A fantastic day in D.C. Met with President Oba...   
...        ...       ...                                                ...   
7417  15-07-16  13:10:00  I loved firing goofball atheist Penn @pennjill...   
7418  15-07-16  10:18:31  I hear @pennjillette show on Broadway is terri...   
7419  15-07-16  10:10:17  Irrelevant clown @KarlRove sweats and shakes n...   
7420  15-07-16   9:44:07  "@HoustonWelder: Donald Trump is one of the se...   
7421  15-07-16   0:21:25  RT @marklevinshow: Trump: Rove is a clown and ...   

      Type   Hashtags      Tweet_Id  \
0     text  ThankAVet  7.970000e+17   
1     text        NaN  7.970000e+17   
2     text        NaN  7.970000e+17   
3     text        NaN  7.970000e+17   
4     text        NaN  7.970000e+17   
...    ...        ...           ...   
7417  text        NaN  6.220000e+17   
7418  text        NaN  6.220000e+17   
7419  text        NaN  6.220000e+17   
7420  text        NaN  6.220000e+17   
7421  link        NaN  6.210000e+17   

                                              Tweet_Url   Likes  Retweets  
0     https://twitter.com/realDonaldTrump/status/797...  127213     41112  
1     https://twitter.com/realDonaldTrump/status/797...  141527     28654  
2     https://twitter.com/realDonaldTrump/status/797...  183729     50039  
3     https://twitter.com/realDonaldTrump/status/796...  214001     67010  
4     https://twitter.com/realDonaldTrump/status/796...  178499     36688  
...                                                 ...     ...       ...  
7417  https://twitter.com/realDonaldTrump/status/621...     953       431  
7418  https://twitter.com/realDonaldTrump/status/621...    1175      1086  
7419  https://twitter.com/realDonaldTrump/status/621...    1494       930  
7420  https://twitter.com/realDonaldTrump/status/621...    1800      1738  
7421  https://twitter.com/realDonaldTrump/status/621...     962       689  

[7422 rows x 9 columns]

### Задание 1

Сделайте чистку данных и удалите все дубликаты.  
Сколько всего уникальных записей в датасете?

In [19]:
df_new = df.drop_duplicates()
df_new.shape


(7375, 9)

7375 уникальных записей

### Задание 2

В датафрейме есть NaN значения.  
Сколько их в столбце хэштегов?

In [20]:
df_drop = df_new.dropna (subset=['Hashtags'])
df_drop.shape




(2031, 9)

5344 или 5391

### Задание 3

Замените **все** значения NaN на ноль.  
Какой наименее популярный тип твита (по количеству записей)?

In [21]:
df_new = df_new.fillna('0')
df_3 = df_new[df_new['Type'] == 'link']
df_3

Date      Time                                         Tweet_Text  \
15    16-11-08  18:23:39  #ElectionDay https://t.co/MXrAxYnTjY https://t...   
17    16-11-08  16:39:36  VOTE TODAY! Go to https://t.co/MXrAxYnTjY to f...   
20    16-11-08   4:32:51  RT @EricTrump: Sean Hannity: If Hillary wins, ...   
22    16-11-08   4:27:18  RT @detroitnews: .@IvankaTrump in Michigan: Th...   
29    16-11-07  21:37:25  America must decide between failed policies or...   
...        ...       ...                                                ...   
7352  15-07-16  18:51:22  Via @FSMtweet: Trump is Right: Illegal Alien C...   
7353  15-07-16  18:45:41  30,000 MACY۪S CUSTOMERS RETALIATE IN SUPPORT O...   
7354  15-07-16  18:37:18  Is this true about Univision and Fusion? Wow!?...   
7355  15-07-16  18:35:32  Via @BreitbartNews by Steve Bannon: ωTIME TO G...   
7374  15-07-16   0:21:25  RT @marklevinshow: Trump: Rove is a clown and ...   

      Type     Hashtags      Tweet_Id  \
15    link  ElectionDay  7.960000e+17   
17    link            0  7.960000e+17   
20    link            0  7.960000e+17   
22    link            0  7.960000e+17   
29    link            0  7.960000e+17   
...    ...          ...           ...   
7352  link            0  6.220000e+17   
7353  link            0  6.220000e+17   
7354  link            0  6.220000e+17   
7355  link            0  6.220000e+17   
7374  link            0  6.210000e+17   

                                              Tweet_Url  Likes  Retweets  
15    https://twitter.com/realDonaldTrump/status/796...  59198     25424  
17    https://twitter.com/realDonaldTrump/status/796...  67787     29756  
20    https://twitter.com/realDonaldTrump/status/795...  20150      9176  
22    https://twitter.com/realDonaldTrump/status/795...  20591      6917  
29    https://twitter.com/realDonaldTrump/status/795...  36663     16108  
...                                                 ...    ...       ...  
7352  https://twitter.com/realDonaldTrump/status/621...    575       356  
7353  https://twitter.com/realDonaldTrump/status/621...    918       425  
7354  https://twitter.com/realDonaldTrump/status/621...    334       176  
7355  https://twitter.com/realDonaldTrump/status/621...    352       201  
7374  https://twitter.com/realDonaldTrump/status/621...    962       689  

[925 rows x 9 columns]

video

### Задание 4

Какой самый популярный хэштег в твитах?

In [30]:
from collections import Counter
hash_ = df_drop['Hashtags'].tolist()
itog = []
for i in hash_:
    i = i.split(';')
    for j in i:
        itog.append(j)
dict = Counter (itog)
res = max (dict, key = dict.get)
print (res)

Trump2016


Trump2016

### Задание 5

Какая медиана количества ретвитов?

In [23]:
df_new["Retweets"].median()


2173.0

2173

### Задание 6

У скольки твитов строго больше 10 000 ретвитов?

In [24]:
df_new[df_new['Retweets'] > 10000]


Date      Time                                         Tweet_Text  \
0     16-11-11  15:26:37  Today we express our deepest gratitude to all ...   
1     16-11-11  13:33:35  Busy day planned in New York. Will soon be mak...   
2     16-11-11  11:14:20  Love the fact that the small groups of protest...   
3     16-11-11   2:19:44  Just had a very open and successful presidenti...   
4     16-11-11   2:10:46  A fantastic day in D.C. Met with President Oba...   
...        ...       ...                                                ...   
4868  15-11-13  21:36:29  They laughed at me when I said to bomb the ISI...   
5623  15-10-14   2:33:13  Notice that illegal immigrants will be given O...   
6249  15-09-12  11:07:38  "@HamishP95: @realDonaldTrump My Dad is thinki...   
6365  15-09-03  16:49:11  Congratulations to Tom Brady on yet another gr...   
6651  15-08-08  12:29:09  So many "politically correct" fools in our cou...   

      Type   Hashtags      Tweet_Id  \
0     text  ThankAVet  7.970000e+17   
1     text          0  7.970000e+17   
2     text          0  7.970000e+17   
3     text          0  7.970000e+17   
4     text          0  7.970000e+17   
...    ...        ...           ...   
4868  text  Trump2016  6.650000e+17   
5623  text  DemDebate  6.540000e+17   
6249  text          0  6.430000e+17   
6365  text          0  6.390000e+17   
6651  text          0  6.300000e+17   

                                              Tweet_Url   Likes  Retweets  
0     https://twitter.com/realDonaldTrump/status/797...  127213     41112  
1     https://twitter.com/realDonaldTrump/status/797...  141527     28654  
2     https://twitter.com/realDonaldTrump/status/797...  183729     50039  
3     https://twitter.com/realDonaldTrump/status/796...  214001     67010  
4     https://twitter.com/realDonaldTrump/status/796...  178499     36688  
...                                                 ...     ...       ...  
4868  https://twitter.com/realDonaldTrump/status/665...   27527     16022  
5623  https://twitter.com/realDonaldTrump/status/654...   26438     17382  
6249  https://twitter.com/realDonaldTrump/status/642...    9077     12681  
6365  https://twitter.com/realDonaldTrump/status/639...   16554     12665  
6651  https://twitter.com/realDonaldTrump/status/629...   21554     12812  

[778 rows x 9 columns]

778

### Задание 7

У скольки твитов строго больше 50 000 лайков с типом 'text'?

In [25]:
df_new[(df_new['Likes'] > 50000) & (df_new['Type'] == 'text')]


Date      Time                                         Tweet_Text  \
0     16-11-11  15:26:37  Today we express our deepest gratitude to all ...   
1     16-11-11  13:33:35  Busy day planned in New York. Will soon be mak...   
2     16-11-11  11:14:20  Love the fact that the small groups of protest...   
3     16-11-11   2:19:44  Just had a very open and successful presidenti...   
4     16-11-11   2:10:46  A fantastic day in D.C. Met with President Oba...   
...        ...       ...                                                ...   
1519  16-07-07   1:34:20  Where is the outrage for this Disney book? Is ...   
1537  16-07-05  15:39:06  FBI director said Crooked Hillary compromised ...   
1538  16-07-05  15:37:23  The system is rigged. General Petraeus got in ...   
1542  16-07-04  22:30:15  Why is President Obama allowed to use Air Forc...   
2158  16-05-05  18:57:30  Happy #CincoDeMayo! The best taco bowls are ma...   

      Type      Hashtags      Tweet_Id  \
0     text     ThankAVet  7.970000e+17   
1     text             0  7.970000e+17   
2     text             0  7.970000e+17   
3     text             0  7.970000e+17   
4     text             0  7.970000e+17   
...    ...           ...           ...   
1519  text        Frozen  7.510000e+17   
1537  text  RiggedSystem  7.500000e+17   
1538  text             0  7.500000e+17   
1542  text             0  7.500000e+17   
2158  text   CincoDeMayo  7.280000e+17   

                                              Tweet_Url   Likes  Retweets  
0     https://twitter.com/realDonaldTrump/status/797...  127213     41112  
1     https://twitter.com/realDonaldTrump/status/797...  141527     28654  
2     https://twitter.com/realDonaldTrump/status/797...  183729     50039  
3     https://twitter.com/realDonaldTrump/status/796...  214001     67010  
4     https://twitter.com/realDonaldTrump/status/796...  178499     36688  
...                                                 ...     ...       ...  
1519  https://twitter.com/realDonaldTrump/status/750...   62396     26208  
1537  https://twitter.com/realDonaldTrump/status/750...   78318     40657  
1538  https://twitter.com/realDonaldTrump/status/750...   56275     22560  
1542  https://twitter.com/realDonaldTrump/status/750...   56272     22076  
2158  https://twitter.com/realDonaldTrump/status/728...   98285     79386  

[142 rows x 9 columns]

142

### Задание 8

Сколько в среднем (математич.ожидание) лайков набирают твиты, содержащие слово "vote"? Округлить до целого числа.

In [26]:
df_8 = df_new[df_new['Tweet_Text'].str.contains('vote')]
df_8['Likes'].mean()

13519.803088803088

13520

### Задание 9

Какая корреляция между лайками и ретвитами? Округлить до 2 знака после запятой.

In [27]:
np.corrcoef(df_new['Likes'], df_new['Retweets'])


array([[1.        , 0.93887991],
       [0.93887991, 1.        ]])

0,94